component/src/main/java/org/wso2/extension/siddhi/io/report/generators/ReportGenerator.java

```diff
@@ -0,0 +1,151 @@
+import net.sf.jasperreports.engine.JasperExportManager;
+import net.sf.jasperreports.engine.JasperFillManager;
+import net.sf.jasperreports.engine.JasperPrint;
+import net.sf.jasperreports.engine.JasperReport;
+import net.sf.jasperreports.engine.data.JRBeanCollectionDataSource;
+import net.sf.jasperreports.engine.design.JasperDesign;
+import net.sf.jasperreports.engine.export.JRCsvExporter;
+import net.sf.jasperreports.engine.export.ooxml.JRXlsxExporter;
+import net.sf.jasperreports.engine.xml.JRXmlLoader;
+import net.sf.jasperreports.export.SimpleExporterInput;
+import net.sf.jasperreports.export.SimpleOutputStreamExporterOutput;
+import net.sf.jasperreports.export.SimpleWriterExporterOutput;
+import org.wso2.extension.siddhi.io.report.util.DynamicLayoutManager;
+import org.wso2.extension.siddhi.io.report.util.ReportConstants;
+import org.wso2.siddhi.core.exception.SiddhiAppRuntimeException;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * This abstract class provides implementation of methods for the report generation.
+ */
+public abstract class ReportGenerator {
+    private Map<String, String> reportProperties = Collections.EMPTY_MAP;
+
+    public ReportGenerator(Map<String, String> reportProperties) {
+        this.reportProperties = reportProperties;
+    }
+
+
+    public void saveReport(JasperPrint jasperPrint, String outputPath) {
+        String fileName = reportProperties.get(ReportConstants.OUTPUT_PATH);
+        File destFile = null;
+        try {
+            switch (reportProperties.get(ReportConstants.OUTPUT_FORMAT).toLowerCase(Locale.ENGLISH)) {
+                case "csv":
+                    fileName += ".csv";
+                    destFile = new File(fileName);
+                    JRCsvExporter csvExporter = new JRCsvExporter();
+                    csvExporter.setExporterInput(new SimpleExporterInput(jasperPrint));
+                    csvExporter.setExporterOutput(new SimpleWriterExporterOutput(destFile));
+                    csvExporter.exportReport();
+                    break;
+                case "excel":
+                    fileName += ".xlsx";
+                    destFile = new File(fileName);
+                    JRXlsxExporter xlsxExporter = new JRXlsxExporter();
+                    xlsxExporter.setExporterInput(new SimpleExporterInput(jasperPrint));
+                    xlsxExporter.setExporterOutput(new SimpleOutputStreamExporterOutput(destFile));
+                    xlsxExporter.exportReport();
+                    break;
+                case "pdf":
+                    outputPath += ".pdf";
+                    JasperExportManager.exportReportToPdfFile(jasperPrint, outputPath);
+                    break;
+                default:
+                    break;
+
+            }
+        } catch (JRException e) {
+            throw new SiddhiAppRuntimeException("Failed to generate the " + reportProperties.get(ReportConstants
```

Comment:
Lets change the exception message to be more informative & descriptive
`Cannot save report + filename + to + outputPath + .`

Reply:
Fixed with 096783872dcb45afcd00b778af443b4647ad246b